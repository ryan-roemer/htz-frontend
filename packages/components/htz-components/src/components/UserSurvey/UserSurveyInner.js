/* global document */
import React, { Component, } from 'react';
import PropTypes from 'prop-types';
import { FelaComponent, } from 'react-fela';

// import { Form,RadioButton,Button } from '@haaretz/htz-components';
import Form from '../Form/Form';
import { RadioButton, } from '../RadioButton/RadioButton';
import TextInput from '../TextInput/TextInput';
import IconClose from '../Icon/icons/IconClose';
import WelcomeCartoon from './WelcomeCartoon';
import Button from '../Button/Button';

import UPDATE_PAGE_COUNT from './UpdatePageCount';
import UPDATE_USER_SURVEY from './UpdateUserSurvey';

import { Mutation, } from '../ApolloBoundary/ApolloBoundary';
import H from '../AutoLevels/H';
import Section from '../AutoLevels/Section';
import {
  slopeWrapper,
  slopeElement,
  contentWrapper,
  closeButton,
  topWrapper,
  surveyTitle,
  welcomeCartoon,
  whatYouThink,
  thankYou,
  forHelpingUs,
  surveyWrapper,
  radioWrapper,
  textArea,
  radioStyle,
  buttonWrapper,
  button,
} from './UserServeyCss';

// import { stylesPropType, } from '../../propTypes/stylesPropType';

class UserSurveyInner extends Component {
  static propTypes = {
    /* User information regarding the survey previus interaction */
    userInfo: PropTypes.shape({
      reactHtzArticleOptIn: PropTypes.bool,
      pageCount: PropTypes.number,
      welcomeScreenViewed: PropTypes.bool,
      articlePageSurvey: PropTypes.shape({
        isOptOut: PropTypes.bool,
        isSubmitted: PropTypes.bool,
        surveySubmission: PropTypes.string,
        surveyAdditionalText: PropTypes.string,
      }),
    }).isRequired,
    /* User basic details */
    user: PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      id: PropTypes.number,
    }).isRequired,
    /* Display the survey */
    hide: PropTypes.bool,
  };

  static defaultProps = {
    hide: false,
  };

  state = {
    didCount: false,
    userAnswer: null,
    showSurvey: !this.props.hide,
    userText: '',
    surveySubmited: false,
  };

  onCheck = e => {
    const newState = this.state;
    newState.userAnswer = e.target.value;
    this.setState(newState);
  };

  onSubmit = (user, setArticlePageSurvey) => {
    if (this.state.userAnswer) {
      setArticlePageSurvey({
        variables: {
          id: user.id,
          isOptOut: false,
          isSubmitted: true,
          surveySubmission: this.state.userAnswer,
          surveyAdditionalText: this.state.userText,
        },
      });
      const newState = this.state;
      newState.surveySubmited = true;
      this.setState(newState);
    }
  };

  onClose = async (user, setArticlePageSurvey) => {
    try {
      await setArticlePageSurvey({
        variables: {
          id: user.id,
          isOptOut: true,
          isSubmitted: false,
          surveySubmission: null,
          surveyAdditionalText: null,
        },
      });
      this.hideSurvey();
    }
    catch (error) {
      this.hideSurvey();
    }
  };

  hideSurvey = () => {
    const newState = this.state;
    newState.showSurvey = false;
    this.setState(newState);
  };

  render() {
    const userOptions = {
      1: 'מרוצה מאוד',
      2: 'מרוצה',
      3: 'לא מרוצה',
    };

    const radios = Object.keys(userOptions).map(option => (
      <FelaComponent
        style={{ display: 'flex', }}
        render={({ className, }) => (
          <RadioButton
            key={option}
            label={userOptions[option]}
            className={className}
            value={option}
            miscStyles={radioStyle}
            onChange={this.onCheck}
            checked={this.state.userAnswer === option}
          />
        )}
      />
    ));

    let surveyContent;

    if (this.state.surveySubmited) {
      surveyContent = (
        <div>
          <FelaComponent tagName="div" style={closeButton}>
            <IconClose onClick={this.onClose} />
          </FelaComponent>
          <FelaComponent style={topWrapper} render="div">
            <FelaComponent
              style={thankYou}
              render={({ className, }) => (
                <H className={className}>
                  {'תודה '}
                  <FelaComponent style={forHelpingUs} render="span">
                    שעזרת לנו להיות טובים יותר
                  </FelaComponent>
                </H>
              )}
            />
            <FelaComponent style={welcomeCartoon} render="div">
              <WelcomeCartoon size="25" />
            </FelaComponent>
          </FelaComponent>
        </div>
      );
    }
    else {
      surveyContent = (
        <div>
          <Mutation mutation={UPDATE_USER_SURVEY}>
            {setArticlePageSurvey => (
              <div>
                <FelaComponent tagName="div" style={closeButton}>
                  <IconClose
                    onClick={() =>
                      this.onClose(this.props.user, setArticlePageSurvey)
                    }
                  />
                </FelaComponent>
                <FelaComponent style={topWrapper} render="div">
                  <FelaComponent
                    style={surveyTitle}
                    render={({ className, }) => (
                      <H className={className}>
                        {`היי ${this.props.user.firstName}, `}
                        <FelaComponent style={whatYouThink} render="span">
                          מה דעתך על עמוד הכתבה החדש שלנו?
                        </FelaComponent>
                      </H>
                    )}
                  />
                  <FelaComponent style={welcomeCartoon} render="div">
                    <WelcomeCartoon size="25" />
                  </FelaComponent>
                </FelaComponent>
                <Form
                  onSubmit={() =>
                    this.onSubmit(this.props.user, setArticlePageSurvey)
                  }
                  render={({ handleSubmit, }) => (
                    <div>
                      <FelaComponent render="div" style={surveyWrapper}>
                        <FelaComponent render="div" style={radioWrapper}>
                          {radios}
                        </FelaComponent>
                        <FelaComponent style={{ flex: '1', }}>
                          <TextInput
                            variant="primaryInverse"
                            label=""
                            isTextArea
                            labelHidden
                            placeholder="בחוויה שלי..."
                            height={[
                              { from: 's', value: 13, },
                              { until: 's', value: 16, },
                            ]}
                            miscStyles={textArea}
                            onChange={evt =>
                              this.setState({ userText: evt.target.value, })
                            }
                            value={this.state.userText}
                          />
                        </FelaComponent>
                      </FelaComponent>
                      <FelaComponent style={buttonWrapper} render="div">
                        <Button
                          variant="primary"
                          boxModel={{ hp: 3, vp: 1, }}
                          miscStyles={button}
                          onClick={handleSubmit}
                        >
                          שלח
                        </Button>
                      </FelaComponent>
                    </div>
                  )}
                />
              </div>
            )}
          </Mutation>
        </div>
      );
    }

    return (
      <FelaComponent
        style={{
          ...slopeWrapper,
          ...{ display: this.state.showSurvey ? 'block' : 'none', },
        }}
        render={({ className, }) => (
          <Section tagName="aside" className={className}>
            {!this.state.didCount ? (
              <Mutation mutation={UPDATE_PAGE_COUNT}>
                {setPageCount => {
                  setPageCount({
                    variables: {
                      id: this.props.user.id,
                      value: this.props.userInfo.pageCount + 1,
                    },
                  });

                  this.setState({
                    didCount: true,
                  });
                  return null;
                }}
              </Mutation>
            ) : null}
            <FelaComponent style={slopeElement} render="div" />
            <FelaComponent style={contentWrapper} render="div">
              {surveyContent}
            </FelaComponent>
          </Section>
        )}
      />
    );
  }
}

export default UserSurveyInner;
