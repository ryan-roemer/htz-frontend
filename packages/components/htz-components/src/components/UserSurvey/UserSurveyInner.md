```jsx
<div dir="rtl">
  <UserSurveyInner
    user={{ firstName: 'Yehuda', lastName: 'Goldner', id: '9037999662' }}
    userInfo={{
      reactHtzArticleOptIn: null,
      pageCount: 3,
      welcomeScreenViewed: null,
      articlePageSurvey: {
        isOptOut: false,
        isSubmitted: false,
        surveySubmission: null,
        surveyAdditionalText: null,
      },
    }}
  />
</div>
```
