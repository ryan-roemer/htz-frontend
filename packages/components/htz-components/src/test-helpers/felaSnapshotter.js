import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
