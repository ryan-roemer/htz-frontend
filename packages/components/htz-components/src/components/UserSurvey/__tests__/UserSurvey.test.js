import React from 'react';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import UserSurveyInner from '../UserSurveyInner';

const user = { firstName: 'Yehuda', lastName: 'Goldner', id: '9037999662', };
const userInfo = {
  reactHtzArticleOptIn: null,
  pageCount: 3,
  welcomeScreenViewed: null,
  articlePageSurvey: {
    isOptOut: false,
    isSubmitted: false,
    surveySubmission: null,
    surveyAdditionalText: null,
  },
};

describe('<UserSurveyInner />', () => {
  it('should render UserSurveyInner currectly', () => {
    const { component, styles, } = felaSnapshotter(
      <UserSurveyInner user={user} userInfo={userInfo} />
    );
    expect(component).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });
});
