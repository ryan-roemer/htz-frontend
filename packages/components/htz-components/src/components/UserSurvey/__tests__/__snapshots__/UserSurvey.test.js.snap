// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSurveyInner /> should render UserSurveyInner currectly 1`] = `
<div>
  <aside
    className="htz-a htz-b htz-c htz-d htz-e"
  >
    <div
      className="htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n htz-o"
    />
    <div
      className="htz-p htz-q htz-r htz-b htz-s htz-t"
    >
      <div />
    </div>
  </aside>
</div>
`;

exports[`<UserSurveyInner /> should render UserSurveyInner currectly 2`] = `".htz-a{overflow:hidden}.htz-b{position:relative}.htz-c{margin-bottom:5rem}.htz-d{margin-top:5rem}.htz-e{display:block}.htz-f{background-image:-webkit-linear-gradient(to bottom, #EBEBEB, #DAE9F2);background-image:-moz-linear-gradient(to bottom, #EBEBEB, #DAE9F2);background-image:linear-gradient(to bottom, #EBEBEB, #DAE9F2)}.htz-g{transform-origin:top left}.htz-h{transform:skewY(10deg)}.htz-i{width:100%}.htz-j{height:100%}.htz-k{position:absolute}.htz-l{pointer-events:none}.htz-m{top:4rem}.htz-p{padding-left:2rem}.htz-q{padding-right:2rem}.htz-r{padding-bottom:2rem}@media (min-width: 37.5em){.htz-n{top:0}.htz-o{transform:skewY(4deg)}.htz-s{padding-left:5rem}.htz-t{padding-right:5rem}}"`;
